ts/nni_manager/core/tuner_command_channel/websocket_channel.ts

```diff
@@ -0,0 +1,146 @@
+ *  The callback to serve WebSocket connection request. Used by REST server module.
+ *  It should only be invoked once, or an error will be raised.
+ *
+ *  Type hint of express-ws is somewhat problematic. Don't want to waste time on it so use `any`.
+ **/
+export function serveWebSocket(ws: WebSocket): void {
+    channelSingleton.setWebSocket(ws);
+}
+
+class WebSocketChannelImpl {
+    private deferredInit: Deferred<void> = new Deferred<void>();
+    private emitter: EventEmitter = new EventEmitter();
+    private heartbeatTimer!: NodeJS.Timer;
+    private serving: boolean = false;
+    private waitingPong: boolean = false;
+    private ws!: WebSocket;
+
+    public setWebSocket(ws: WebSocket): void {
+        if (this.ws !== undefined) {
+            logger.error('A second client is trying to connect');
+            ws.close(4030, 'Already serving a tuner.');
+            return;
+        }
+
+        logger.debug('Connected.');
+        this.serving = true;
+
+        this.ws = ws;
+        ws.on('close', () => { this.handleError(new Error('tuner_command_channel: Tuner closed connection')); });
+        ws.on('error', this.handleError.bind(this));
+        ws.on('message', this.receive.bind(this));
+        ws.on('pong', () => { this.waitingPong = false; });
+
+        this.heartbeatTimer = setInterval(this.heartbeat.bind(this), heartbeatInterval);
+        this.deferredInit.resolve();
+    }
+
+    public init(): Promise<void> {
+        logger.debug(this.ws === undefined ? 'Waiting connection...' : 'Initialized.');
+        return this.deferredInit.promise;
+    }
+
+    public async shutdown(): Promise<void> {
+        if (this.ws === undefined) {
+            return;
+        }
+        clearInterval(this.heartbeatTimer);
+        this.serving = false;
+        this.emitter.removeAllListeners();
+    }
+
+    public sendCommand(command: string): void {
+        assert.ok(this.ws !== undefined);
+
+        logger.debug('Sending', command);
+        this.ws.send(command);
+
+        if (this.ws.bufferedAmount > command.length + 1000) {
+            logger.warning('Sending too fast! Try to reduce the frequency of intermediate results.');
+        }
+    }
+
+    public onCommand(callback: (command: string) => void): void {
+        this.emitter.on('command', callback);
+    }
+
+    public onError(callback: (error: Error) => void): void {
+        this.emitter.on('error', callback);
+    }
+
+    private heartbeat(): void {
+        if (this.waitingPong) {
+            this.ws.terminate();  // this will trigger "close" event
+            this.handleError(new Error('tuner_command_channel: Tuner loses responsive'));
+        }
+
+        this.waitingPong = true;
+        this.ws.ping();
+    }
+
+    private receive(data: Buffer, _isBinary: boolean): void {
+        logger.debug('Received', data);
+        this.emitter.emit('command', data.toString());
+    }
+
+    private handleError(error: Error): void {
+        if (!this.serving) {
```

Comment:
If this is not serving any more (due to some glitch or whatever reason), should we set `this.ws` to undefined?

In that way, maybe tuner could set a request later, and re-setup a ws, pretending that nothing happened.

Reply:
This field is only used for one case here:
An `error` event will followed the by a `close` event, and in normal cases `close` will trigger on error handler. `serving` will ensure that the handler is not invoked twice.
If we set `this.ws` to `undefined` before the `close` event, I don't know if there will be problems. It requires experiments and is not an easy work. So I don't want to deal with it now.
Anyway the OS guarantees loopback connection is always reliable.